---
title: Promise
tags: 小总结
categories: Promise
date: 2020-04-26
---

# 为什么要用Promise[](https://www.bilibili.com/video/BV1MJ41197Eu?p=18)

1.指定回调函数的方式更加灵活

- 旧的：必须在启动异步任务前指定
- promise：启动异步任务 => 返回promise对象 => 给promise对象绑定回调函数（甚至可以在异步结束后指定）

2.支持链式调用个，可以解决回调地狱问题

- 回调地狱：回调函数嵌套调用，外部回调函数异步执行的结果是嵌套的回调函数执行的条件
- 回调地狱的缺点：不便于阅读、不便于异常处理
- 解决方案：promise链式调用
- 终极解决方案：async/await

# 如何使用Promise

## API

1.Promise构造函数：Promise(executor){}

- (1)excutor函数：执行器(resolve,reject)=>{}
- (2)resolve函数：内部定义成功时调用的函数 value=>{}
- (3)reject函数：内部定义失败时调用的函数 resaon=>{}
- 说明：excutor会在Promise内部立即同步回调，异步操作在执行器中执行

2.Promise.prototype.then 方法：(onResolved,onRejected)=>{}

- (1)onResolved函数：成功的回调函数(value)=>{}
- (2)onRejected函数：失败的回调函数(reason)={}
- 说明：指定用于得到成功value的成功回调和用于得到失败reason的失败回调
            返回一个心的promise对象

3.Promise.prototype.catch方法：(onRejected)=>{}

- (1)onRejected函数：失败的回调函数(reason)={}
- 说明：then()的语法糖，相当于:then(undefined,onRejected)

4.Promise.resolve方法：(value)={}

- (1)value：成功的数据或promise对象
- 说明：返回一个成功/失败的promise对象

5.Promise.reject方法：(reason)={}

- (1)reason：失败的数据或promise对象
- 说明：返回一个失败的promise对象

6。Promise.all方法：(promises)={}

- (1)promises：包含n个promise的数组
- 说明：返回一个新的promise，只有所有的promise都成功才成功，只要有一个失败直接失败

7.Promise.rece方法：(promises)={}

- (1)promises：包含n个promise
- 个完成的promise的结果状态就是最终的结果状态

# promise的几个关键问题

1.如何改变promise的状态

- resolve(value)：如果当前是pending都会改变为resolved
- reject(value)：如果当前是pending都会改变为rejected
- 抛出异常：如果当前是pending都会改变为rejected

2.1个promise指定多个成功/失败回调函数，都会调用吗？

- 当promise改变为对应状态时都会调用

3.改变promise状态和指定回馈函数谁先谁后？

- 都有可能，正常情况下是先指定回调再改变状态，但也可以先改变状态再指定回调
- 如何先改变状态再指定回调？
  1、在执行器中直接调用resolve()/reject()
  2、延迟更长时间才调用then()
- 什么时候才能的得到数据
  1、如果先指定的回调，那当状态发生改变时，回调函数就会调用，得到数据
  2、如果先改变的状态，那当指定回调时，回调函数就会调用，得到数据

4.promise.then()返回的新promise的结果状态由什么决定？

- 简单表达：由then()指定的回调函数执行的 结果决定
- 详细表达：
  1、如果抛出异常，新promise变为rejected，reason为抛出的异常
  2、如果返回的是非promise的任意值，新promise变为resolved，value为返回的值
  3、如果返回的是另一个新promise，此promise的结果就会成为新promise的结果

5.promise如何串联多个操作任务？

- promise的then()返回一个新的promise，可以开成then()的链式调用
- 通过then的链式调用串连多个同步/异步任务

6.promise异常传透？

- 当使用promise的then链式调用时，可以在最后指定失败的回调，
- 前面任何系统出了异常，都会传回最后失败的回调中处理

7.中断promise链？

- 当使用promise的then链式调用时，在中间中断，不再调用后面的回调函数
- 办法：在回调函数中返回pending状态的promise对象